util/src/main/java/tech/pegasys/teku/util/events/ObservableValue.java

```diff
@@ -0,0 +1,137 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.util.events;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A value holder class which notifies subscribers on value updates
+ *
+ * <p>The key feature of this class is that upon subscription a new subscriber is always notified on
+ * the current value (if the value exist) and this is performed in a thread-safe manner such that
+ * {@link #set(Object)} and {@link #subscribe(Consumer)} methods can be safely called from different
+ * threads
+ *
+ * <p>All subscribers are guaranteed:
+ *
+ * <ul>
+ *   <li>to be always notified on the latest value
+ *   <li>to be notified on each value just once
+ *   <li>to be notified in the right order if the {@link #set(Object)} is invoked on the same thread
+ * </ul>
+ *
+ * Initially the holder has no value so added subscribers are not notified upon subscription
+ *
+ * @param <C> Value type
+ */
+public class ObservableValue<C> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private static final class Subscription<C> {
+    private final Consumer<C> subscriber;
+    private final long subscriptionId;
+
+    public Subscription(Consumer<C> subscriber, long subscriptionId) {
+      this.subscriber = subscriber;
+      this.subscriptionId = subscriptionId;
+    }
+
+    public Consumer<C> getSubscriber() {
+      return subscriber;
+    }
+
+    public long getSubscriptionId() {
+      return subscriptionId;
+    }
+  }
+
+  private long idCounter = 0;
+  private final List<Subscription<C>> subscriptions = new CopyOnWriteArrayList<>();
+  private C curValue;
+  private final boolean suppressCallbackExceptions;
+
+  /**
+   * Creates instance
+   *
+   * @param suppressCallbackExceptions if true then any exceptions thrown from a subscriber update
+   *     callback are just printed to the log
+   */
+  public ObservableValue(boolean suppressCallbackExceptions) {
+    this.suppressCallbackExceptions = suppressCallbackExceptions;
+  }
+
+  /**
+   * Subscribe to value update notification
+   *
+   * <p>New subscriber is notified on the latest value if the value exist
+   *
+   * @return subscription ID to be used for {@link #unsubscribe(long)}
+   */
+  public synchronized long subscribe(Consumer<C> subscriber) {
```

Comment:
It's probably worth creating a specific interface rather than using the generic `Consumer` here.  Makes it easier to understand where values are coming from in the classes that implement it.  Something like `ValueObserver` with an `onValueChanged` method.

Reply:
Yes, sounds good 👍